[FLINK-14019][python] add support for managing environment and dependencies of Python UDF in Flink Python API

WeiZhong94 opened the pull request:
## What is the purpose of the change

*This pull request adds support for managing environment and dependencies of Python UDF in Flink Python API.*


## Brief change log

  - *Add new APIs in flink-python to support upload users' python environment and dependencies to cluster.*
  - *Refactor the operator layer of flink-python to support different type of running mode(docker mode, process mode, etc).*
  - *Add ProcessEnvironmentManager to support managing environment and dependencies of Python UDF worker when running in process mode.*


## Verifying this change

This change is already covered by newly added tests, such as *ProcessEnvironmentManagerTest*, *PythonDependencyManagerTest*, and so on.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (python docs)

WeiZhong94 commented:
@flinkbot run travis

WeiZhong94 commented:
@hequn8128 Thanks you for the the notification! This failure did not occur in my own travis, but it happened twice in flink-ci travis. So I think it may be due to a cache error or other random factors. I''m trying to push something to refresh the cache and fix this.

hequn8128 commented on flink-python/pyflink/common/dependency_manager.py:
```diff
@@ -0,0 +1,121 @@
+
+    PYTHON_FILE_PREFIX = "python_file"
+    PYTHON_REQUIREMENTS_PREFIX = "python_requirements_list_file"
+    PYTHON_REQUIREMENTS_DIR_PREFIX = "python_requirements_dir"
```
Change to python_requirements_cached_dir? To make it consistent with `def set_python_requirements(self, requirements_list_file, requirements_cached_dir=None):`

And maybe change python_requirements_list_file to python_requirements_file?

hequn8128 commented on flink-python/pyflink/table/table_environment.py:
```diff
@@ -583,6 +585,114 @@ def register_function(self, name, function):
+        self._dependency_manager.add_python_file(file_path)
+        self._dependency_manager.transmit_parameters_to_jvm(self.get_config().get_configuration())
+
+    def set_python_requirements(self, requirements_list_file, requirements_cached_dir=None):
```
Rename requirements_list_file to requirements_file?

hequn8128 commented on flink-python/src/main/java/org/apache/flink/python/PythonDependencyManager.java:
```diff
@@ -0,0 +1,183 @@
+	private String requirementsFilePath;
+	private String requirementsDirPath;
+	private String pythonExec;
+	private Map<String, String> archives;
```
Use @Nullable annotation for these class members.

hequn8128 commented on flink-python/src/main/java/org/apache/flink/python/PythonDependencyManager.java:
```diff
@@ -0,0 +1,183 @@
+	 * 	       TableEnvironment#add_python_file() or command option "-pyfs".
+	 * 	       Key is local absolute path and value is origin file name.
+	 */
+	public Map<String, String> getFilesInPythonPath() {
```
Always use Optional to return nullable values. See more details in [Code Style and Quality Guide](https://flink.apache.org/contributing/code-style-and-quality-java.html)

hequn8128 commented on flink-python/src/main/java/org/apache/flink/python/PythonDependencyManager.java:
```diff
@@ -0,0 +1,183 @@
+		String pythonExec = null;
+		if (dependencyMetaData.get(PYTHON_EXEC) != null) {
+			pythonExec = dependencyMetaData.get(PYTHON_EXEC);
+		}
```
It is equivalent to `String pythonExec=dependencyMetaData.get(PYTHON_EXEC);`

hequn8128 commented on flink-python/src/main/java/org/apache/flink/python/PythonDependencyManager.java:
```diff
@@ -0,0 +1,183 @@
+		throws IOException {
+		Map<String, String> filesPathToFilesName = new HashMap<>();
+		if (dependencyMetaData.get(PYTHON_FILE_MAP) != null) {
+			ObjectMapper mapper = new ObjectMapper();
```
Put mapper as a method variable and avoid define it multi times?

WeiZhong94 commented:
@hequn8128 Thanks for your review! I have addressed your comments in the latest commits.

hequn8128 commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,317 @@
+			}
+			Path src = FileSystems.getDefault().getPath(entry.getKey());
+			try {
+				if (testCopy) {
```
Remove this test logic?

hequn8128 commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,317 @@
+				}
+				Files.createSymbolicLink(target, src);
+			} catch (IOException e) {
+				FileUtils.copy(new org.apache.flink.core.fs.Path(src.toUri()),
```
Add log info if create symbolic link failed, e.g., LOG.warn().

hequn8128 commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,317 @@
+ * function runner is configured to run python UDF in process mode.
+ */
+@Internal
+public class ProcessEnvironmentManager implements PythonEnvironmentManager {
```
public final?

WeiZhong94 commented on flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractPythonFunctionOperator.java:
```diff
@@ -75,6 +80,7 @@
 	 * A timer that finishes the current bundle after a fixed amount of time.
 	 */
 	private transient ScheduledFuture<?> checkFinishBundleTimer;
+	protected PythonEnvironmentManager environmentManager;
```
I agree that it is unnecessary to add it here. 
But the creation of environment manager relies on the information provided by flink operators, such as pythonEnv, temp directory path, dependency manager, and these data is no use for Runner except creating environment manager. Maybe we can initialize the environment manager in the base operator and pass to Runner immediately in open method? 
If move the creation to Runner it will introduce additional concrete class dependency (ProcessEnvironmentManager and DockerEnvironmentManager in the future) of python environment managers.

WeiZhong94 commented on flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractPythonFunctionOperator.java:
```diff
@@ -227,6 +251,9 @@ public void processWatermark(Watermark mark) throws Exception {
 	 */
 	public abstract PythonFunctionRunner<IN> createPythonFunctionRunner();
 
+	public abstract PythonEnvironmentManager createPythonEnvironmentManager(
```
In the latest commit I refactored here and the situation here is much different now... Could you review here again?

WeiZhong94 commented:
@hequn8128 Thanks for your review again! I have addressed most of your comments and reply some of them because I think it may need some discussion there. Please take a look :).

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,353 @@
+
+	private static void appendToPythonPath(Map<String, String> env, String libs) {
+		String systemPythonPath = env.get("PYTHONPATH");
+		if (systemPythonPath != null && !systemPythonPath.isEmpty()) {
```
What happens if systemPythonPath is empty?

WeiZhong94 replied:
If we not check empty there will be an extra path separator at the end of PYTHONPATH.

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,353 @@
+			.collect(Collectors.joining(File.pathSeparator));
+		appendToPythonPath(systemEnv, internalLibs);
+
+		// if the log directory exists, transmit it to the runner
```
Do you want to mean "transmit it to the worker"?

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,353 @@
+	}
+
+	@VisibleForTesting
+	void prepareEnvironment() {
```
Can this method be removed?

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,353 @@
+	 * @throws IllegalArgumentException
+	 */
+	private void prepareWorkingDir() throws IOException, IllegalArgumentException {
+
```
remove empty line

dianfu commented on flink-python/src/main/java/org/apache/flink/python/PythonDependencyManager.java:
```diff
@@ -0,0 +1,183 @@
+	public static final String PYTHON_ARCHIVES_MAP = "python.environment.archive-map";
+	public static final String PYTHON_EXEC = "python.environment.exec";
+
+	@Nonnull private Map<String, String> filesInPythonPath;
```
final?

dianfu commented on flink-python/src/main/java/org/apache/flink/python/util/UnzipUtil.java:
```diff
@@ -0,0 +1,108 @@
+		}
+	}
+
+	private static boolean isMacOrUnix() {
```
Is it possible to use the util OperatingSystem?

dianfu commented on flink-python/src/main/java/org/apache/flink/python/util/UnzipUtil.java:
```diff
@@ -0,0 +1,108 @@
+/**
+ * Utils used to extract zip files and try to restore the origin permissions of files.
+ */
+public class UnzipUtil {
```
Is there third party library which could be reused as this requirement seems very common.

WeiZhong94 replied:
Of course there is third party library which could be reused, but it will introduce additional dependencies. The apache-common-compress library is introduced in flink-core so reuse the classes in it won't introduce any new dependencies.

WeiZhong94 commented on flink-python/pyflink/fn_execution/boot.py:
```diff
@@ -45,105 +46,195 @@
+        line_continuation = None
+        with open(requirements_file_path, "r") as f:
+            while True:
+                text_line = f.readline()
```
IMHO we should give users a chance to install and upgrade the pip itself or other package-installation-related packages (e.g. wheel), and installing the packages line by line allows users to do that.

WeiZhong94 commented on flink-python/pyflink/table/table_environment.py:
```diff
@@ -715,6 +718,111 @@ def create_temporary_view(self, view_path, table):
+        :param target_dir: The target dir name that the archive file extracted to.
+        :type target_dir: str
+        """
+        self._dependency_manager.add_python_archive(archive_path, target_dir)
```
To check whether a file is zip-format we must read its actual content. If user specifies a DFS url the check will introduce additional IO. Suffix checking is also hard because there are too many file formats are actually zip format. Maybe rewriting the doc string in detail to tell users what file format are actually supported is a better choice?

WeiZhong94 commented on flink-python/pyflink/common/dependency_manager.py:
```diff
@@ -0,0 +1,111 @@
+                    break
+            self._parameters.remove_config(parameter_name)
+
+    def _register_file(self, parameter_name, file_path, set_parameters=False):
```
Actually it is config key so I rename it to "config_key".

WeiZhong94 commented on flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/PythonScalarFunctionOperator.java:
```diff
@@ -106,15 +107,16 @@ public void emitResults() {
-	public PythonFunctionRunner<Row> createPythonFunctionRunner(FnDataReceiver<Row> resultReceiver) {
+	public PythonFunctionRunner<Row> createPythonFunctionRunner(
+			FnDataReceiver<Row> resultReceiver,
+			PythonEnvironmentManager pythonEnvironmentManager) {
```
PythonEnvironmentManager is created from the information provided by AbstractPythonFunctionOperator and used in PythonFunctionRunner so maybe it is not a best choice that let it become a field of AbstractPythonFunctionOperator. But we can still remove it from here to keep the interface unchanged.

WeiZhong94 commented:
@dianfu Thanks for your review! I have addressed your comments. Please take a look again. :)

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,361 @@
+	public String createRetrievalToken() throws IOException {
+		File retrievalToken = new File(baseDirectory,
+			"retrieval_token_" + UUID.randomUUID().toString() + ".json");
+		if (!retrievalToken.getParentFile().exists() && !retrievalToken.getParentFile().mkdir()) {
```
It seems that this check is not necessary any more as the parent directory will be created when creating the baseDirectory.

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,361 @@
+			LOG.info("requirements install directory of python worker: {}", requirementsDirectory);
+		}
+
+		// Transmit the path of python interpreter to bootstrap script.
```
bootstrap script -> the boot script

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,361 @@
+		// Transmit the path of python interpreter to bootstrap script.
+		if (dependencyManager.getPythonExec().isPresent()) {
+			systemEnv.put("python", dependencyManager.getPythonExec().get());
+			LOG.info("user defined python interpreter path: {}", dependencyManager.getPythonExec());
```
The first word should be upper case. Pls also check other places.

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,361 @@
+	}
+
+	@VisibleForTesting
+	String getbaseDirectory() {
```
getbaseDirectory -> getBaseDirectory

dianfu commented on flink-python/src/main/java/org/apache/flink/python/PythonDependencyManager.java:
```diff
@@ -0,0 +1,183 @@
+	 *                 Key is local absolute path and value is target directory name.
+	 */
+	public PythonDependencyManager(
+		Map<String, String> filesInPythonPath,
```
filesInPythonPath -> pythonFiles?

dianfu commented on flink-python/src/main/java/org/apache/flink/python/PythonDependencyManager.java:
```diff
@@ -0,0 +1,183 @@
+	 * @return The PythonDependencyManager object that contains whole information of python dependency.
+	 */
+	public static PythonDependencyManager create(
+		Map<String, String> dependencyMetaData, DistributedCache distributedCache)
```
Rename dependencyMetaData to globalJobParameters? globalJobParameters is more understandable in Flink.

dianfu commented on flink-python/src/main/java/org/apache/flink/python/PythonDependencyManager.java:
```diff
@@ -0,0 +1,183 @@
+ * This class is used to parse the information of python dependency and environment management
+ * stored in GlobalJobParameters. The parse result will be used to create PythonEnvironmentManager.
+ */
+@Internal
```
I don't see much value of this class. Does it make sense to you remove it?

WeiZhong94 replied:
In the future the DockerPythonEnvironmentManager will reuse this class, so IMO keeping it here is a better choice.

WeiZhong94 replied:
I have rename it to PythonDependencyInfo. Maybe this name makes more sense to this class.

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,361 @@
+ * and create ProcessEnvironment object of Beam Fn API. It will be created if the python
+ * function runner is configured to run python UDF in process mode.
+ */
+@Internal
```
Move PythonEnvironmentManager and ProcessEnvironmentManager to package o.a.f.python.env?

dianfu commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,361 @@
+import java.util.stream.Collectors;
+
+/**
+ * The ProcessEnvironmentManager used to prepare the working dir of python UDF worker
```
Rename ProcessEnvironmentManager to ProcessPythonEnvironmentManager?

dianfu commented on flink-python/src/main/java/org/apache/flink/python/AbstractPythonFunctionRunner.java:
```diff
@@ -198,19 +174,14 @@ public void open() throws Exception {
 		Struct pipelineOptions = PipelineOptionsTranslation.toProto(portableOptions);
 
 		jobBundleFactory = createJobBundleFactory(pipelineOptions);
+		// The creation of stageBundleFactory depends on the initialized environment manager.
```
Move `environmentManager.open();` before portableOptions statement?

dianfu commented on flink-python/src/main/java/org/apache/flink/streaming/api/operators/python/AbstractPythonFunctionOperator.java:
```diff
@@ -275,4 +287,27 @@ private void invokeFinishBundle() throws Exception {
+				.getTaskManagerInfo()
+				.getConfiguration()
+				.getString(ConfigConstants.TASK_MANAGER_LOG_PATH_KEY, System.getProperty("log.file"));
+			String logDirectory = taskManagerLogFile == null ? null : new File(taskManagerLogFile).getParent();
```
In which case taskManagerLogFile is null?

WeiZhong94 replied:
e.g. running IT cases in IDE.

dianfu commented on flink-python/src/main/java/org/apache/flink/table/runtime/operators/python/AbstractPythonScalarFunctionOperator.java:
```diff
@@ -145,13 +148,21 @@ public void processElement(StreamRecord<IN> element) throws Exception {
+	}
+
+	@Override
+	public PythonFunctionRunner<IN> createPythonFunctionRunner() throws IOException {
```
The IOException could be removed if createBaseDirectory is done in the open method of ProcessEnvironmentManager.

WeiZhong94 replied:
The PythonDependencyInfo#create still throws IOException so we can not remove it currently.

WeiZhong94 commented on flink-python/bin/pyflink-udf-runner.sh:
```diff
@@ -23,9 +23,22 @@ if [[ "$FLINK_TESTING" = "1" ]]; then
+    python="python"
+fi
+
+if [[ "$_PYTHON_WORKING_DIR" != "" ]]; then
```
If we change it as above, sometimes users will not be able to access archives through relative paths because the python working directory is not set as expected.

WeiZhong94 commented on flink-python/pyflink/table/table_environment.py:
```diff
@@ -883,8 +987,8 @@ def __init__(self, j_tenv, is_blink_planner):
         self._j_tenv = j_tenv
         super(StreamTableEnvironment, self).__init__(j_tenv, is_blink_planner)
 
-    def _get_execution_config(self, filename, schema):
```
Dependency manager needs execution environment to register files, so we need to implement a "_get_j_env" method to get it, which can completely replace this method. The old implementation also has a few problem and it's a internal method, so I removed it.

WeiZhong94 commented on flink-python/src/main/java/org/apache/flink/python/ProcessEnvironmentManager.java:
```diff
@@ -0,0 +1,361 @@
+
+	private static final Logger LOG = LoggerFactory.getLogger(ProcessEnvironmentManager.class);
+
+	public static final String PYTHON_REQUIREMENTS_FILE = "_PYTHON_REQUIREMENTS_FILE";
```
The public declaration can not be removed because they are used in python tests via py4j.

WeiZhong94 commented:
@dianfu Thanks for your review again! I have addressed your comments. Please take a look.

dianfu commented on flink-python/src/main/java/org/apache/flink/python/env/ProcessPythonEnvironmentManager.java:
```diff
@@ -0,0 +1,361 @@
+		Map<String, String> systemEnv = new HashMap<>(this.systemEnv);
+
+		// add pyflink, py4j and cloudpickle to PYTHONPATH
+		String internalLibs = Arrays.stream(ResourceUtil.PYTHON_BASIC_DEPENDENCIES)
```
Rename PYTHON_BASIC_DEPENDENCIES to BUILT_IN_PYTHON_DEPENDENCIES?

dianfu commented on flink-python/src/test/java/org/apache/flink/python/env/ProcessPythonEnvironmentManagerTest.java:
```diff
@@ -0,0 +1,448 @@
+		}
+	}
+
+	private static String computeSHA1Checksum(File file) throws IOException, NoSuchAlgorithmException {
```
Use org.apache.commons.io.FileUtils.contentEquals()?

WeiZhong94 commented on flink-python/src/main/java/org/apache/flink/python/util/UnzipUtil.java:
```diff
@@ -0,0 +1,111 @@
+				File file;
+				if (entry.isDirectory()) {
+					file = new File(targetPath, entry.getName());
+					if (!file.exists()) {
```
If the entry is extracted after its underlying files, the file.exists will be true.

WeiZhong94 commented on flink-python/src/main/java/org/apache/flink/python/util/UnzipUtil.java:
```diff
@@ -0,0 +1,111 @@
+/**
+ * Utils used to extract zip files and try to restore the origin permissions of files.
+ */
+public class UnzipUtil {
```
The ProcessPythonEnvironmentManagerTest has covered this util class.

WeiZhong94 commented:
@dianfu Thanks for your review! I have addressed your comments in the latest commit.

dianfu commented:
@WeiZhong94 Thanks for the update. LGTM.

hequn8128 commented:
It seems the failure of Travis does not relate to this pr.
